### PR TITLE
feat(api): POST /ms-trophy-challenge-completed

### LINKED ISSUE #51430: POST /ms-trophy-challenge-completed

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes https://github.com/freeCodeCamp/freeCodeCamp/issues/51430

I didn't have chance to finish this 100% before my vacation (sorry!), but the functionality is done. The only changes I have left to do are to finish the changes to the tests. I separated out the calls to Microsoft's api into some helper functions, which I could then mock instead of fetch... but I haven't finished.

I'll do that when I get back, but if people want to merge before then feel free to make any changes you want.

### CODE REVIEW

**Comment by moT01** [comment]
The linked PR above made some changes to this endpoint. They should probably be added here.

File: api/src/routes/challenge.ts [hunk 1]
```diff
@@ -424,5 +426,104 @@ export const challengeRoutes: FastifyPluginCallbackTypebox = (
...
+        if (msTrophyStatus.type === 'error') {
+          return reply.code(403).send(msTrophyStatus);
+        }
+
+        const user = await fastify.prisma.user.findUniqueOrThrow({
+          where: { id: req.session.user.id },
+          select: { completedChallenges: true, progressTimestamps: true }
+        });
+
+        const { completedChallenges } = user;
```
Comment (moT01) [inline api/src/routes/challenge.ts:484]: Curious why we didn't use the `updateUserChallengeData` function here

**Review by moT01 — COMMENTED** [review]
This mostly LGTM :tada: I left one thought. We need to add the changes I mentioned still, too.

File: api/src/routes/challenge.ts (hunk 1 above)
Comment (ojeytonwilliams) [reply api/src/routes/challenge.ts:484]: That function has to handle too many special cases and is quite confusing. I'd rather be a little less DRY for now, while we're building the functionality, and try to refactor later if needed.

**Comment by ojeytonwilliams** [comment]
@moT01 I've incorporated the response for when the user has no/private achievements.

**Review by moT01 — APPROVED** [review]

**Review by Sembauke — APPROVED** [review]

**Pull request merged**